File: pravega/src/main/java/akka/stream/alpakka/pravega/javadsl/Pravega.java
/*
 * Copyright (C) since 2016 Lightbend Inc. <https://www.lightbend.com>
 */

package akka.stream.alpakka.pravega.javadsl;

import akka.Done;
import akka.NotUsed;
import akka.annotation.ApiMayChange;
import akka.stream.alpakka.pravega.PravegaEvent;

import akka.stream.alpakka.pravega.WriterSettings;
import akka.stream.alpakka.pravega.PravegaReaderGroupManager;
import akka.stream.alpakka.pravega.*;
import akka.stream.javadsl.Flow;
import akka.stream.javadsl.Keep;
import akka.stream.javadsl.Sink;
import akka.stream.javadsl.Source;

import io.pravega.client.ClientConfig;
import io.pravega.client.stream.ReaderGroup;
import scala.compat.java8.FutureConverters;
import java.util.concurrent.CompletionStage;

import akka.stream.alpakka.pravega.impl.PravegaFlow;
import akka.stream.alpakka.pravega.impl.PravegaSource;

@ApiMayChange
public class Pravega {

  public static PravegaReaderGroupManager readerGroup(String scope, ClientConfig clientConfig) {

    return new PravegaReaderGroupManager(scope, clientConfig);
  }
  /**
   * Messages are read from a Pravega stream.
   *
   * <p>Materialized value is a future which completes to Done as soon as Pravega reader in open.
   */
  public static <V> Source<PravegaEvent<V>, CompletionStage<Done>> source(
      ReaderGroup readerGroup, ReaderSettings<V> readerSettings) {
    return Source.fromGraph(new PravegaSource<>(readerGroup, readerSettings))
        .mapMaterializedValue(FutureConverters::<Done>toJava);
  }

  /** Incoming messages are written to Pravega stream and emitted unchanged. */
  public static <V> Flow<V, V, NotUsed> flow(
      String scope, String streamName, WriterSettings<V> writerSettings) {
    return Flow.fromGraph(new PravegaFlow<>(scope, streamName, writerSettings));
  }
  /** Incoming messages are written to Pravega. */
  public static <V> Sink<V, CompletionStage<Done>> sink(
      String scope, String streamName, WriterSettings<V> writerSettings) {
    return flow(scope, streamName, writerSettings).toMat(Sink.ignore(), Keep.right());
  }
}

File: pravega/src/main/scala/akka/stream/alpakka/pravega/scaladsl/Pravega.scala
/*
 * Copyright (C) since 2016 Lightbend Inc. <https://www.lightbend.com>
 */

package akka.stream.alpakka.pravega.scaladsl
import akka.annotation.ApiMayChange
import akka.stream.scaladsl.{Flow, Keep, Sink, Source}
import akka.{Done, NotUsed}
import akka.stream.alpakka.pravega.impl.{PravegaFlow, PravegaSource}
import akka.stream.alpakka.pravega.{PravegaEvent, PravegaReaderGroupManager, ReaderSettings, WriterSettings}
import io.pravega.client.ClientConfig
import io.pravega.client.stream.ReaderGroup

import scala.concurrent.Future

@ApiMayChange
object Pravega {

  def readerGroupManager(scope: String, clientConfig: ClientConfig) = new PravegaReaderGroupManager(scope, clientConfig)

  /**
   * Messages are read from a Pravega stream.
   *
   * Materialized value is a [[Future]] which completes to [[Done]] as soon as the Pravega reader is open.
   */
  def source[A](readerGroup: ReaderGroup, readerSettings: ReaderSettings[A]): Source[PravegaEvent[A], Future[Done]] =
    Source.fromGraph(new PravegaSource(readerGroup, readerSettings))

  /**
   * Incoming messages are written to Pravega stream and emitted unchanged.
   */
  def flow[A](scope: String, streamName: String, writerSettings: WriterSettings[A]): Flow[A, A, NotUsed] =
    Flow.fromGraph(new PravegaFlow(scope, streamName, writerSettings))

  /**
   * Incoming messages are written to Pravega.
   */
  def sink[A](scope: String, streamName: String, writerSettings: WriterSettings[A]): Sink[A, Future[Done]] =
    Flow[A].via(flow(scope, streamName, writerSettings)).toMat(Sink.ignore)(Keep.right)

}
